### PR TITLE
Add Neighbor and NeighborVerlet classes

### LINKED ISSUE #38: Add Neighbor class
Add back `Neighbor` class for increased flexibility, initially with `VerletList` subclass

### PR DESCRIPTION
**Pull request by streeve**
Add flexibility to neighbor list generation and remove from Force classes (previous code duplication).

Currently just base `Neighbor` and `NeighborVerlet`, but this is set up to add `NeighborTree` (ECP-copa/Cabana#208) if it's converted to a class.

Closes #38

### CODE REVIEW

**Comment by streeve** [comment]
Just a note - the last commit fix was caught by ECP-copa/Cabana#210

File: src/cabanamd.cpp [hunk 1]
```diff
@@ -102,33 +102,107 @@ void CabanaMD::init( int argc, char *argv[] )
     // Create Binning class: linked cell bin sort
     binning = new Binning( system );
 
+    // Create Neighbor class: create neighbor list
+    bool half_neigh = input->force_iteration_type == FORCE_ITER_NEIGH_HALF;
+    if ( input->neighbor_type == NEIGH_VERLET_2D )
```
Comment (sslattery) [inline src/cabanamd.cpp]: This is a nice factory interface. Unfortunately one of the consequences of using templates to get performance is at some point a pattern like this is needed to expand all of the possible types to allow a selection to be made based on user input.

File: src/cabanamd.cpp [hunk 2]
```diff
@@ -102,33 +102,107 @@ void CabanaMD::init( int argc, char *argv[] )
...
+        }
+        else
+        {
+            if ( input->neighbor_type == NEIGH_VERLET_2D )
+                force = new ForceLJ<t_verletlist_full_2D>( system );
+            else if ( input->neighbor_type == NEIGH_VERLET_CSR )
+                force = new ForceLJ<t_verletlist_full_CSR>( system );
+        }
+    }
+#ifdef CabanaMD_ENABLE_NNP
```
Comment (sslattery) [inline src/cabanamd.cpp:114]: Is `NNP` optional because of an optional library dependence?

File: src/cabanamd.cpp [hunk 3]
```diff
@@ -102,33 +102,107 @@ void CabanaMD::init( int argc, char *argv[] )
...
+    else if ( input->force_type == FORCE_NNP )
     {
+        if ( half_neigh )
+            throw std::runtime_error( "Half neighbor list not implemented "
```
Comment (sslattery) [inline src/cabanamd.cpp:119]: Nice - good to check for bad input combinations.

File: src/cabanamd.cpp [hunk 4]
```diff
@@ -152,12 +226,12 @@ void CabanaMD::init( int argc, char *argv[] )
     comm->exchange_halo();
 
     // Compute atom neighbors
-    force->create_neigh_list( system );
+    neighbor->create( system );
```
Comment (sslattery) [inline src/cabanamd.cpp:200]: So now the neighbor list is split from forces? Much better design.

File: src/cabanamd.cpp [hunk 5]
```diff
@@ -292,7 +366,7 @@ void CabanaMD::run( int nsteps )
         if ( step % input->thermo_rate == 0 )
         {
             T_FLOAT T = temp.compute( system );
-            T_FLOAT PE = pote.compute( system, force ) / system->N;
+            T_FLOAT PE = pote.compute( system, force, neighbor ) / system->N;
```
Comment (sslattery) [inline src/cabanamd.cpp:344]: for all these you could do `auto` as the return type - the compiler will automatically deduce the return type from the function call.

File: src/neighbor.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+template <class t_neigh_list>
+class NeighborVerlet : public Neighbor
+{
+  public:
+    t_neigh_list list;
+
+    void init( T_X_FLOAT neigh_cut, bool half_neigh );
+    void create( System *system );
+
+    t_neigh_list &get();
```
Comment (sslattery) [inline src/neighbor.h]: When we add the tree class how will the type `t_neigh_list` be resolved? Will both the verlet and tree lists return different types? If so this lends itself to the force functions being templated on a neighbor list type.

File: src/neighbor.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    virtual const char *name();
+};
+
+template <class t_neigh_list>
+class NeighborVerlet : public Neighbor
+{
+  public:
+    t_neigh_list list;
+
+    void init( T_X_FLOAT neigh_cut, bool half_neigh );
```
Comment (sslattery) [inline src/neighbor.h]: mark these as `override` to indicate they are implementations of base class functions.

File: src/neighbor.h [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+{
+  public:
+    T_X_FLOAT neigh_cut;
+    bool half_neigh;
+
+    virtual void init( T_X_FLOAT neigh_cut, bool half_neigh );
+    virtual void create( System *system );
+
+    template <class t_neigh_list>
+    t_neigh_list &get();
```
Comment (sslattery) [inline src/neighbor.h]: Does the base class have an implementation of this function?

File: src/neighbor.h [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+#include <system.h>
+#include <types.h>
+
+class Neighbor
+{
+  public:
+    T_X_FLOAT neigh_cut;
+    bool half_neigh;
+
+    virtual void init( T_X_FLOAT neigh_cut, bool half_neigh );
```
Comment (sslattery) [inline src/neighbor.h]: If these functions don't have default implementations in the base class declare them pure virtual by adding a `= 0` at the end.

File: src/neighbor.h (hunk 6 above)
Comment (sslattery) [reply src/neighbor.h]: Ultimately I think we want to converge on a single Cabana neighbor list data type (independent of the construction method such as Verlet or Tree). Then you would return that type here and this function can also be pure virtual in the base class. In the short term I think this makes the most sense and we move the Cabana neighbor list traits interface into an interface in that simple class. In the future we will explore computing forces as we traverse the tree which would be a different paradigm.

**Review by sslattery — CHANGES_REQUESTED** [review]
Mostly some clean up in the new `Neighbor` interface

File: src/cabanamd.cpp (hunk 2 above)
Comment (streeve) [reply src/cabanamd.cpp:114]: Yeah, that's the only reason

File: src/neighbor.h (hunk 6 above)
Comment (streeve) [reply src/neighbor.h]: Agreed. Following from the layout PR, I plan to template both Neighbor and Force on the neighbor type, which would work fine with separate Cabana list types. But I think this current approach could work with Cabana list types that are covariant - when I tested adding the Tree class I saw that you're right, this doesn't work with multiple child classes

**Comment by streeve** [comment]
I merged in changes from #30 and extended that to the new `Neighbor` class here

File: src/cabanamd.cpp (hunk 1 above)
Comment (streeve) [reply src/cabanamd.cpp]: Agreed. This is mainly moved to `mdfactory.h` in #30 and continued here

File: src/neighbor.h (hunk 6 above)
Comment (streeve) [reply src/neighbor.h]: This is taken care of, as you suggested, with a template over `Neighbor` types

File: src/neighbor.h (hunk 7 above)
Comment (streeve) [reply src/neighbor.h]: Done

File: src/neighbor.h (hunk 8 above)
Comment (streeve) [reply src/neighbor.h]: It does now - much simpler with the templates

File: src/neighbor.h (hunk 9 above)
Comment (streeve) [reply src/neighbor.h]: Done

**Comment by streeve** [comment]
Closing in favor of updated and rebased #43

**Pull request closed without merging**